[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Primagon/CSc44800_Assignment_04/blob/main/CSc44800_Assignment4.ipynb)

In [1]:
## all the imports and pips
import numpy as np
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.neighbors import KNeighborsClassifier

C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\feature_extraction\image.py:167: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  dtype=np.int):
C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\linear_model\least_angle.py:30: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.o

# Part 1. Loading the dataset

In [2]:
iris = datasets.load_iris()

C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\datasets\base.py:241: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  target = np.empty((n_samples,), dtype=np.int)
C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\datasets\base.py:245: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, 

C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\datasets\base.py:245: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  target[i] = np.asarray(ir[-1], dtype=np.int)
C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\datasets\base.py:245: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, c

In [3]:
# Load the dataset (load remotely, not locally)

# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

iris_df = pd.DataFrame(iris.data, columns = iris.feature_names)

iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                 5.1               3.5                1.4               0.2
1                 4.9               3.0                1.4               0.2
2                 4.7               3.2                1.3               0.2
3                 4.6               3.1                1.5               0.2
4                 5.0               3.6                1.4               0.2
5                 5.4               3.9                1.7               0.4
6                 4.6               3.4                1.4               0.3
7                 5.0               3.4                1.5               0.2
8                 4.4               2.9                1.4               0.2
9                 4.9               3.1                1.5               0.1
10                5.4               3.7                1.5               0.2
11                4.8               3.4                1.6               0.2
12                4.8               3.0                1.4               0.1
13                4.3               3.0                1.1               0.1
14                5.8               4.0                1.2               0.2

In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


## About the dataset

The features for the iris dataset are:
<br>
sepal length
<br>
sepal width
<br>
petal length
<br>
petal width
<br>
<br>
The label for the iris dataset are:
<br>
class
<br>
<br>
The model splits the label into 3 subclasses. In the iris dataset, they are given numerical values of 0, 1 and 2. 
<br>
0 would be for Iris-setosa, 1 for Iris-versicolor, and 2 for Iris-virginica.

# Part 2: Split the dataset into train and test

In [5]:
# Take the dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

features = iris.data
features

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [6]:
labels = iris.target
labels

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [7]:
x_train, x_test, y_train, y_test = train_test_split(features, labels, test_size = 0.1)

# Part 3: Logistic Regression

In [8]:
# i. Use sklearn to train a LogisticRegression model on the training set

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for Logistic regression model, what does the score measure?

# iv. Extract the coefficents and intercepts for the boundary line(s)

#Part i
reg = LogisticRegression(random_state=0).fit(x_train, y_train)
reg.predict(x_test)

C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


array([0, 1, 0, 2, 2, 0, 1, 0, 2, 2, 2, 1, 0, 0, 0])

## Sample datapoint to be used to predict the probabilities for each possible class for all future parts

In [9]:
sample = np.array([[2, 2, 1, 3]])

## Part 3.ii) For a sample datapoint, predict the probabilities for each possible class

In [10]:
reg.predict_proba(sample)

array([[0.18898209, 0.00597079, 0.80504712]])

## Part 3.iii) Report the score for Logistic regression model

In [11]:
reg.score(x_test, y_test)

1.0

With a score of 1.0, my model predicts roughly around 100% correct for classification using a logistic regression.

## Part 3.iv) The coefficients and intercepts of the boundary line(s)

In [12]:
reg.coef_

array([[ 0.40376761,  1.41820986, -2.21915824, -0.99138504],
       [ 0.40662179, -1.58352258,  0.61568144, -1.43958342],
       [-1.66107498, -1.4646626 ,  2.38259196,  2.48493729]])

In [13]:
reg.intercept_

array([ 0.25635787,  1.07341206, -1.17902337])

## Part 4: Support Vector Machine

In [14]:
# i. Use sklearn to train a Support Vector Classifier on the training set

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for the SVM, what does the score measure?

#Part i
clf = svm.SVC(kernel = 'linear', probability=True)
clf.fit(x_train, y_train)
clf.predict(x_test)

array([0, 1, 0, 2, 2, 0, 1, 0, 2, 2, 2, 1, 0, 0, 0])

## Part 4.ii) For a sample datapoint, predict the probabilities for each possible class

In [15]:
clf.predict_proba(sample)

array([[0.58508979, 0.35728919, 0.05762101]])

## Part 4.iii) Report the score for SVM

In [16]:
clf.score(x_test, y_test)

1.0

With a score of 1.0 my model predicts roughly around 100% correct for classification using SVM

## Part 5: Neural Network

In [17]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for the Neural Network, what does the score measure?

# iv: Experiment with different options for the neural network, report on your best configuration 
#(the highest score I was able to achieve was 0.8666)

#Part i
scaler = StandardScaler()
scaler.fit(x_train)
x_train = scaler.transform(x_train)
x_test = scaler.transform(x_test)
print(x_train[:3])

[[-0.53512929  1.58663342 -1.31972183 -1.35954565]
 [ 1.58597904 -0.09051264  1.22959448  1.19146722]
 [-0.28558713 -0.09051264  0.18669236  0.11735654]]


In [18]:
mlp = MLPClassifier(hidden_layer_sizes=(10, 5), max_iter=1000)

# let's fit the training data to our model
mlp.fit(x_train, y_train)
mlp.predict(x_test)

array([0, 1, 0, 2, 2, 0, 1, 0, 2, 1, 2, 1, 0, 0, 0])

## Part 5.ii) For a sample datapoint, predict the probabilities for each possible class

In [19]:
mlp.predict_proba(sample)

array([[2.41426083e-06, 7.01629818e-03, 9.92981288e-01]])

## Part 5.iii) Report the score for Neural Network (MLP Classifier)

In [20]:
mlp.score(x_test, y_test)

0.9333333333333333

With a score of 0.9333, my model predicts roughly around 93.3% correct for classification using Neural Network (MLP Classifier) with hyperparameters hidden_layer_sizes and max_iter changed. Everything else is default.

## Part 5.iv) Experimentations with different hyperparameters

In [21]:
#MLP with only learning rate changed
mlp2 = MLPClassifier(learning_rate = "adaptive")
mlp2.fit(x_train, y_train)
mlp2.predict(x_test)

C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


array([0, 1, 0, 2, 2, 0, 1, 0, 2, 1, 2, 1, 0, 0, 0])

In [22]:
mlp2.score(x_test, y_test)

0.9333333333333333

With a score of 0.9333, my model predicts roughly around 93.3% correct for classification using Neural Network (MLP Classifier) with hyperparameters learning_rate changed. Everything else is default.

In [23]:
#MLP with learning rate and hidden layer changed
mlp3 = MLPClassifier(hidden_layer_sizes=(100, 10), max_iter=1000, learning_rate = "invscaling") 
mlp3.fit(x_train, y_train)
mlp3.predict(x_test)

array([0, 1, 0, 2, 2, 0, 1, 0, 2, 1, 2, 1, 0, 0, 0])

In [24]:
mlp3.score(x_test, y_test)

0.9333333333333333

With a score of 0.9333, my model predicts roughly around 93.3% correct for classification using Neural Network (MLP Classifier) with hyperparameters hidden_layer_sizes, learning_rate, and max_iter changed. Everything else is default.

## Part 6: K-Nearest Neighbors

In [25]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

# ii. For a sample datapoint, predict the probabilities for each possible class

# iii. Report on the score for kNN, what does the score measure?

#Part i
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(x_train, y_train)
classifier.predict(x_test)

C:\Users\kenne\Anaconda3\lib\site-packages\sklearn\neighbors\base.py:908: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  self._y = np.empty(y.shape, dtype=np.int)


array([0, 1, 0, 2, 2, 0, 1, 0, 2, 1, 2, 1, 0, 0, 0])

## Part 6.ii) For a sample datapoint, predict the probabilities for each possible class

In [26]:
classifier.predict_proba(sample)

array([[0., 0., 1.]])

## Part 6.iii) Report the score for K-Nearest Neighbors

In [27]:
classifier.score(x_test, y_test)

0.9333333333333333

With a score of 0.9333, my model predicts roughly around 93.3% correct for classification using K-Nearest Neighbors.

# Part 7: Conclusions and takeaways

The models that performed the best on the iris dataset were the Logistic Regression Model and the SVM model because they have the highest score out of all the models. 
<br><br>
I can see why the Logistic Regression Model performed as one of the best because of how the model handles the data. In Logistic Regression, the input values (x) are combined linearly using weights or coefficient values (Beta) to predict an output value (y). 
<br><br>
The same can be said for SVM model. In the SVM model, the algorithm creates line(s) or hyperplane(s) which separates the data into classes. Then it would separate the data based on the lines/hyperplane(s) to help classify which class the datapoint should be classified as. 
<br><br>
What surprised me about the exercise is how certain the K-Neighest Neighbor model was for predicting the probability of each class for the sample datapoint I made. It was 100% certain that the sample datapoint is in class 2. Then another surpising thing about the exercise was that for every other model, the score was 0.9333. Another surprise was that sklearn had their own inbuilt datasets. It was my first time loading data from sklearn and learning how to call the features and labels compared to how I did it in assignment 3.